src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php

```diff
@@ -249,4 +275,92 @@ private function assertProductCanBeUpdated(
             }
         }
     }
+
+    /**
+     * @param Cart $cart
+     * @param Order $order
+     * @param OrderInvoice|null $invoice
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function updateCartRules(Cart $cart, Order $order, ?OrderInvoice $invoice = null)
+    {
+        $computingPrecision = Context::getContext()->getComputingPrecision();
+
+        $cart = $this->syncCartOrderCartRules($order, $cart);
+        Context::getContext()->cart = $cart;
+        CartRule::autoAddToCart();
+        CartRule::autoRemoveFromCart();
+
+        $orderCartRulesData = $order->getCartRules();
+        foreach ($orderCartRulesData as $orderCartRuleData) {
+            $orderCartRule = new OrderCartRule((int) $orderCartRuleData['id_order_cart_rule']);
+            $idCartRule = (int) $orderCartRule->id_cart_rule;
+            $cartRule = new CartRule($idCartRule);
+            $values = [
+                'tax_incl' => \Tools::ps_round($cartRule->getContextualValue(true), $computingPrecision),
+                'tax_excl' => \Tools::ps_round($cartRule->getContextualValue(false), $computingPrecision),
+            ];
+
+            if (
+                ($values['tax_incl'] !== \Tools::ps_round($orderCartRule->value, $computingPrecision)) ||
+                ($values['tax_excl'] !== \Tools::ps_round($orderCartRule->value_tax_excl, $computingPrecision))
+            ) {
+                $orderCartRule->value = $values['tax_incl'];
+                $orderCartRule->value_tax_excl = $values['tax_excl'];
+
+                $orderCartRule->update();
+            }
+        }
+
+        $order = $this->orderAmountUpdater->update($order, $cart, $invoice && !empty($invoice->id));
+        $order->update();
+    }
+
+    /**
+     * @param Order $order
+     * @param Cart $cart
+     *
+     * @return Cart
+     */
+    private function syncCartOrderCartRules(Order $order, Cart $cart): Cart
+    {
+        $orderCartRulesData = $order->getCartRules();
+        $orderCartRulesIds = array_map(
+            function (array $orderCartRuleData) { return (int) $orderCartRuleData['id_cart_rule']; },
+            $orderCartRulesData
+        );
+
+        $cartRules = $cart->getCartRules();
+        $cartRulesIds = array_map(
+            function (array $cartRule) { return (int) $cartRule['id_cart_rule']; },
+            $cartRules
+        );
+
+        sort($orderCartRulesIds);
+        sort($cartRulesIds);
+
+        $difference = array_diff($orderCartRulesIds, $cartRulesIds);
+
+        if (count($orderCartRulesIds) > count($cartRulesIds)) {
+            foreach ($difference as $cartRuleId) {
+                $cart->addCartRule($cartRuleId);
+            }
+        } else {
+            foreach ($difference as $cartRuleId) {
```

Comment:
I do not have the case where some rules are on the cart but not on the order. Maybe I need to remove this until we need it ?

Reply:
I dont know either if it could happen.

3 possible solutions
- ignore it, we dont handle it because we think it never happens
- ignore it, we dont handle it because we have no idea what should happen and we hope "do nothing" is the right thing to do
- throw a RuntimeException "should never happen" here

Reply:
Solution 3 is brutal but I like it. We can't miss it if it happens